### PR TITLE
Allow to store logs in files

### PR DESCRIPTION
**Pull request by qwerty287**
https://github.com/woodpecker-ci/woodpecker/discussions/2278

### CODE REVIEW

**Comment by zc-devs** [comment]
> Do we need native support for something like S3 or can we just ask users to mount it directly into the FS?

> in object storage

Native S3.

1. I believe, environments exist, where you cannot mount arbitrary FUSE (AWS EKS, Windows, I guess).
2. I'm not sure if FUSE S3 can patch a file. If can't, then this is not well from performance and maybe costs perspective.

---

Sadly, there is no description in the issue, but I'm guessing it is related to [growing Database](https://github.com/woodpecker-ci/woodpecker/pull/3506#issuecomment-2022159070). If so, [tiering across storage mediums](https://min.io/product/automated-data-tiering-lifecycle-management) could be implemented.
1. Some bunch of data stores in DB.
2. After some retention condition old data archives to S3 (#1068).

**Comment by anbraten** [comment]
Could we try to merge #3722 first or do you think they wouldn't conflict that much?

File: server/services/log/file/file.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		err = os.MkdirAll(base, 0600)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return logStore{base: base}, nil
+}
+
+func (l logStore) filePath(id int64) string {
+	return filepath.Join(l.base, strconv.Itoa(int(id))+".json")
```
Comment (anbraten) [inline server/services/log/file/file.go:33]:
```suggestion
	return filepath.Join(l.base, fmt.Sprintf("%d.json", id))
```

Comment (anbraten) [inline server/services/log/file/file.go:33]: Would it make sense to include the repo id in the path to improve grouping in the folder?

File: server/services/log/file/file.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	file, err := os.ReadFile(l.filePath(step.ID))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	var entries []*model.LogEntry
+	for _, j := range bytes.Split(file, []byte("\n")) {
```
Comment (anbraten) [inline server/services/log/file/file.go:46]: Would be better to use a scanner instead I think.

File: cmd/server/server.go [hunk 3]
```diff
@@ -276,6 +276,11 @@ func setupEvilGlobals(c *cli.Context, s store.Store) error {
 	}
 	server.Config.Services.Manager = serviceManager
 
+	server.Config.Services.LogStore, err = setupLogStore(c, s)
+	if err != nil {
+		return fmt.Errorf("could not setup log store manager: %w", err)
```
Comment (anbraten) [inline cmd/server/server.go:281]:
```suggestion
		return fmt.Errorf("could not setup log store: %w", err)
```

File: cmd/server/setup.go [hunk 4]
```diff
@@ -154,3 +156,12 @@ func setupMetrics(g *errgroup.Group, _store store.Store) {
 		}
 	})
 }
+
+func setupLogStore(c *cli.Context, s store.Store) (logService.Service, error) {
+	switch c.String("log-store") {
+	case "file":
+		return file.NewLogStore(c.String("log-store-file-base"))
```
Comment (anbraten) [inline cmd/server/setup.go:163]:
```suggestion
		return file.NewLogStore(c.String("log-store-file-path"))
```

File: docs/docs/30-administration/10-server-config.md [hunk 5]
```diff
@@ -543,6 +543,18 @@ Enable the Swagger UI for API documentation.
...
+
+> Default: `database`
+
+Where to store logs. Possible values: `database` or `file`.
+
+### `WOODPECKER_LOG_STORE_FILE_PATH`
+
+> Default empty
+
+Directory to store logs if [`WOODPECKER_LOG_STORE`](#woodpecker_log_store) is `file`.
```
Comment (anbraten) [inline docs/docs/30-administration/10-server-config.md:556]:
```suggestion
Directory to store logs in if [`WOODPECKER_LOG_STORE`](#woodpecker_log_store) is `file`.
```

**Review by anbraten — APPROVED** [review]

**Pull request merged**